draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1216 @@
+:ref:`draft/celery-5-high-level-architecture:Services` in order to track
+changes in the environment it is running on.
+
+.. admonition:: Example
+
+  We have a task which requires 8GB of memory to complete.
+  The worker runs a service which constantly monitors the system's available
+  memory.
+  If there is not enough memory it changes the task's health check to the
+  **Unhealthy** state.
+
+If a task or a service that is part of a health check fails unexpectedly it
+is ignored and an error message is logged.
+
+Celery provides many types of health checks in order to verify that it can
+operate without any issues.
+
+Users may implement their own health checks in addition to the built-in health
+checks.
+
+Some health checks are specific to the worker they are executing on.
+Therefore, their state is stored in-memory in the worker.
+
+Other health checks are global to all or a group of workers.
+As such, their state is stored externally.
+
+If the state storage for health checks is not provided, these health checks
+are disabled.
+
+Health Checks can be associated with tasks in order to ensure that they are
+likely to succeed. Multiple Health Check failures may trigger
+a :term:`Circuit Breaker` which will prevent the task from running for a period
+of time or automatically mark it as failed.
+
+Each Health Check declares its possible states.
+Sometimes it makes sense to try to execute a task anyway even if the
+health check occasionally fails.
+
+.. admonition:: Example
+
+  A health check that verifies whether we can send a HTTP request to an endpoint
+  has multiple states.
+
+  The health check performs an
+  `OPTIONS <https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS>`_
+  HTTP request to that endpoint and expects it to respond within the specified
+  timeout.
+
+  The health check is in a **Healthy** state if all the following conditions are
+  met:
+
+  * The DNS server is responding within the specified time limit and is
+    resolving the address correctly.
+  * The TLS certificates are valid and the connection is secure.
+  * The Intrusion Detection System reports that the network is secure.
+  * The HTTP method we're about to use is listed in the OPTIONS response's
+    `ALLOW <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Allow>`_
+    header.
+  * The content type we're about to format the request in is listed in the
+    OPTIONS response's
+    `ACCEPT <https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept>`_
+    header.
+  * The OPTIONS request responds within the specified time limits.
+  * The OPTIONS request responds with
+    `200 OK <https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/200>`_
+    status.
+
+  In addition, the actual request performed in the task must also stand in the
+  aforementioned conditions. Otherwise, the health check will change it's state.
+
+  The health check can be in an **Insecure** state if one or more of the
```

Comment:
Not sure if this responsibility belongs to Celery.

Reply:
This is just an example.
Your health check can have as many states as you want.

It's your choice on how to respond to them.